x/sequencer/keeper/msg_server_arweave_block.go

```diff
@@ -0,0 +1,117 @@
+package keeper
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/warp-contracts/sequencer/x/sequencer/types"
+)
+
+func (k msgServer) ArweaveBlock(goCtx context.Context, msg *types.MsgArweaveBlock) (*types.MsgArweaveBlockResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if err := k.setNewArweaveBlockInfo(ctx, msg); err != nil {
+		return nil, err
+	}
+
+	return &types.MsgArweaveBlockResponse{}, nil
+}
+
+func (k msgServer) setNewArweaveBlockInfo(ctx sdk.Context, block *types.MsgArweaveBlock) error {
+	var newBlockInfo = &types.ArweaveBlockInfo{
+		Height:    block.BlockInfo.Height,
+		Timestamp: block.BlockInfo.Timestamp,
+		Hash:      block.BlockInfo.Hash,
+	}
+
+	if err := k.checkBlockIsOldEnough(ctx, newBlockInfo); err != nil {
+		return err
+	}
+
+	if err := k.compareBlockWithPreviousOne(ctx, newBlockInfo); err != nil {
+		return err
+	}
+
+	if err := k.compareWithNextBlockAndRemove(ctx, block); err != nil {
+		return err
+	}
+
+	k.SetLastArweaveBlock(ctx, *newBlockInfo)
+	return nil
+}
+
+func (k msgServer) checkBlockIsOldEnough(ctx sdk.Context, newBlockInfo *types.ArweaveBlockInfo) error {
+	arweaveBlockTimestamp := time.Unix(int64(newBlockInfo.Timestamp), 0)
+	cosmosBlockTimestamp := ctx.BlockHeader().Time
+
+	if !types.CheckArweaveBlockIsOldEnough(ctx, newBlockInfo) {
+		return errors.Wrapf(types.ErrArweaveBlockTimestampMismatch,
+			"The timestamp of the Arweave block (%s) should be one hour earlier than the Cosmos block (%s)",
+			arweaveBlockTimestamp.UTC(), cosmosBlockTimestamp.UTC())
+	}
+	return nil
+}
+
+func (k msgServer) compareBlockWithPreviousOne(ctx sdk.Context, newValue *types.ArweaveBlockInfo) error {
+	oldValue, isFound := k.GetLastArweaveBlock(ctx)
+
+	if isFound {
```

Comment:
reverse the condition. The less code in an if block the better (imho)

Reply:
done